### ISSUE TITLE
Developer Guide should include instructions on writing/running tests for certbot-auto

### ISSUE THREAD

**Issue by void-elf** [certbot/certbot#4947]
They should go at the bottom of this section: https://certbot.eff.org/docs/contributing.html#updating-certbot-auto-and-letsencrypt-auto

**Comment by bmw** [comment]
I assigned this to myself to make sure Erica had what she needed to work on #2208. I talked to her in person and we came up with a plan for testing that issue.

Personally, I'd like to see testing for `certbot-auto` redone. There are a couple approaches to testing certbot-auto (`letsencrypt-auto-source/tests` and test farm tests) and both are complicated with a number of gotchas. Because of this, tests for `certbot-auto` are often not written.

Rather than trying to explain the complexity of the current approach, I'd rather eliminate it by making certbot-auto tests easier to write and then documenting that. I'm going to try and make this happen. If we don't end up redoing these tests soon, documenting what we currently have should still be done, but personally I'm going to focus on simplifying what we currently have first.

**Comment by bmw** [comment]
Here's a `certbot-auto` testing brain dump. If I forget anything please add it to this issue or if anyone working on this stuff has any questions, let me know. Some of the pain points I describe below should be taken into account when redesigning these tests.

There are two kinds of `certbot-auto` tests. First are the tests in `letsencrypt-auto-source/tests/auto_test.py` and the other are test farm tests in `tests/letstest/scripts/`. Both of these methods test `certbot-auto` as a whole rather than testing individual pieces.

#### letsencrypt-auto-source/tests/auto_test.py

This modifies pieces of `certbot-auto` to make testing easier. It uses a dummy PyPI server, GitHub server, signatures, and has `certbot-auto` install a simple Python package that spits out a version number rather than the real version of Certbot and all of its Python dependencies. All OS dependencies are still installed (if expected by the test).

For each test, a new directory is used for `certbot-auto`'s installation, but any installed OS packages persist between tests. We have Dockerfiles to run these tests on different systems in `letsencrypt-auto-source`. Feel free to add new ones to the `master` branch and then configure it to be run in the `.travis.yml` in `test-everything` branch as appropriate.

#### tests/letstest/scripts/

All of our "test farm" tests are found here but the ones specifically for testing `certbot-auto` that we currently run are `test_leauto_upgrades.sh` and `test_letsencrypt_auto_certonly_standalone.sh`. Feel free to modify these (as long as you're not deleting existing valuable tests) or add new ones.

There are a number of pain points for writing test farm tests in general such as it takes minutes to get the results of the tests after making changes and sometimes the connection to the EC2 instances failing, but I won't go into more detail about that kind of stuff here. Instead, there are a few gotchas specifically for `certbot-auto` tests you should be aware of.

First, you probably want to include `--no-self-upgrade` on the command line to prevent `certbot-auto` from downgrading itself from the version in `master` to the latest released version. Second, if you've added a flag to `certbot-auto` that Certbot needs to quietly accept, you can't test this without modifying `certbot-auto` because the latest released version of Certbot is pinned in the script. Third, make sure you're testing `letsencrypt-auto-source/letsencrypt-auto` as the versions of `*-auto` in the root of our repository only change during releases.

#### general gotchas

In addition to the test farm test specific pain points, there are a few additional general ones. We don't currently test a real `certbot-auto` upgrade where phase 1 of the latest released (or older) version downloads the new version of the script and reruns itself starting at phase 2. This could potentially be simulated by running an old `certbot-auto`, replacing it with the new one, and running the new script with expected CLI flags (including `--le-auto-phase2`) but this hasn't been done or tested.

Partially because we're testing the script as a whole, there's not a great way to test OS specific functionality. This isn't done at all currently in `letsencrypt-auto-source/tests/auto_test.py` and in the test farm tests it is done by using the `OS_TYPE` environment variable, but this only gives you a high level description of the system. See `tests/letstest/scripts/test_apache2` for an example of this.

Another minor gotcha is because the name of the script being tested is `letsencrypt-auto`, non-interactive mode is assumed for backwards compatibility reasons. You can change the name of the script to literally anything else to work around this.

This is long and a bit unorganized, but I hope it helps.

cc @ohemorange

**Comment by ynasser** [comment]
Two things to maybe add to the documentation based on my experience working with these tests a few months ago:

- it would be helpful if we could document which tests are known to fail randomly/all the time (unless if those have been fixed or removed?), so no one wastes time thinking it's their fault when they first start running them.
- I think this part should be in bold: "**include --no-self-upgrade on the command line to prevent certbot-auto from downgrading itself from the version in master to the latest released version.**" I tripped over this so many times!

**Comment by erikrose** [comment]
I think we can kill 2 birds...

* Tests are inadequate, inefficient, and split up.
* certbot-auto is difficult to reason about and change (see https://github.com/certbot/certbot/pull/5392#issuecomment-356721814 and some of my preceding comments).

...with one stone.

When I wrote auto_tests.py, certbot-auto behaved pretty much the same on all OSes; the only difference was the selection of bootstrapper. So integration tests were arguably fine and certainly better than nothing. Since then, however, we've accumulated more system-specific branches, leading to a profusion of global variables, often with uncertain semantics and scopes of effect. Also, branchy functions like `DeterminePythonVersion()` cry out for unit tests, both for speed and for pointing more precisely to failures.

Let's make certbot-auto itself more modular—that is, looser-coupled. Let's tighten those variable scopes and rigorously define the behavior of individual functions, then unit-test the ones that have a lot of branches. This will also make the whole mess easier to change. I'm not saying it will be easy, given the limitations of the Bourne shell, but I bet we can figure something out. We may even be able to move the latter phases into a different language. I'm happy to help sketch up solutions.
